### PR TITLE
Gk/tp strong scaling clean

### PR DESCRIPTION
**Pull request by greg-kwasniewski1**
# [https://github.com/NVIDIA/TensorRT-LLM/issues/4308][bug] TP > num_kv_heads fix

## Description
Current TP strategy always distributes weight matrix either row-wise or column-wise using the entire WORLD_SIZE. For weight matrices such as w_k of shape [num_kv_heads * head_dim, embd], if TP > num_kv_heads , this results in distributing the individual heads, which leads to incorrect attention computation.

This PR fixes it by passing optional parameter `min_local_size`, derived from `model.config` to the `column_row_shard` logic. Then it calculates maximum number of distributed groups `num_groups = world_size * min_local_shape / t.shape[d]`.
If `world_size > max_split_size`, then the sharder groups ranks into `num_groups` groups, each rank in the group
sharing (replicating) the same slice of the weight tensor.

## Test Coverage

Added new class GQABlock to `test_graph_sharding.py` unittest, which explicitly specifies q_proj, k_proj, v_proj, with `num_kv_heads < num_heads`.

### CODE REVIEW

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py [hunk 1]
```diff
@@ -14,6 +14,9 @@
 from tensorrt_llm._torch.auto_deploy.transformations.library import column_row_shard
 from tensorrt_llm._torch.auto_deploy.utils.node_utils import is_op
 
+num_heads = 4
+num_key_value_heads = 2
+
```
Comment (lucaslie) [inline tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:19]: Looks like a classic case for pytest test parameters. You can look at other unit tests how it is being used

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py [hunk 2]
```diff
@@ -28,6 +31,112 @@ def forward(self, x):
...
+    #     x1 = x[..., ::2]
+    #     x2 = x[..., 1::2]
+
+    #     # Apply RoPE rotation
+    #     rotated_x = torch.cat([
+    #         x1 * cos_freqs[..., ::2] - x2 * sin_freqs[..., 1::2],
+    #         x1 * sin_freqs[..., ::2] + x2 * cos_freqs[..., 1::2]
+    #     ], dim=-1)
+
+    #     return rotated_x
```
Comment (lucaslie) [inline tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:98]: We can skip rope for now and you can delete this code

File: tensorrt_llm/_torch/auto_deploy/transformations/library/sharding.py [hunk 3]
```diff
@@ -85,6 +92,15 @@ def _insert_sharded_matmul(
     def split_tensor(
         t: torch.Tensor, d: int = dim, r: int = rank, ws: int = world_size
     ) -> torch.Tensor:
+        # The local tensor shape has to be divisible by min_local_shape
+        max_split_size = t.shape[d] // min_local_shape
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/sharding.py:96]: The `split_tensor` function is used as a load hook outside the context of the `_insert_sharded_matmul`. So you have to "bind" `min_local_shape` from the outer context to this local function definition so it can later retrieve the correct value

```suggestion
    def split_tensor(
        t: torch.Tensor, d: int = dim, r: int = rank, ws: int = world_size, min_d_shape: int = min_local_shape
    ) -> torch.Tensor:
        # The local tensor shape has to be divisible by min_d_shape
        max_split_size = t.shape[d] // min_d_shape
```

File: tensorrt_llm/_torch/auto_deploy/transformations/library/sharding.py [hunk 4]
```diff
@@ -85,6 +92,15 @@ def _insert_sharded_matmul(
...
+        if ws > max_split_size:
+            num_groups = math.ceil(ws / max_split_size)
+            ad_logger.warning(
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transformations/library/sharding.py:99]:
```suggestion
            ad_logger.debug(
```
Looks more like a case for debug log than a warning. This behavior is unaccounted for - so nothing to warn about

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py [hunk 5]
```diff
@@ -28,6 +31,112 @@ def forward(self, x):
         return self.linear2(y)
 
 
+class GQABlock(nn.Module):
```
Comment (lucaslie) [inline tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:34]: There is a proper GQA block defined in `tests/unittest/_torch/auto_deploy/_utils_test/_model_test_utils.py`

We can just reuse it

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py (hunk 1 above)
Comment (greg-kwasniewski1) [reply tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:19]: I'm not sure if we need to treat these as tunable, varying test parameters like `@pytest.mark.parametrize("num_heads", [4, 8])`, do you think it's necessary? I see more analogies to hardcoded model parameters that I see throughout the tests:
```
# test_ep_sharding.py
    hidden_size = 32
    intermediate_size = 16
# test_graph_sharding.py
    batch_size = 4
    
    (...)
```
I thought that it would unnecessarily spawn too many tests, and the corner-case we want to test is that `num_heads > num_key_value_heads` is small. Or do you prefer to parameterize it?

Comment (greg-kwasniewski1) [reply tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:20]: @lucaslie  what do you think?

Comment (lucaslie) [reply tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_graph_sharding.py:20]: Seems we should at least test the following cases though:

* num_kv_heads // world_size >= 1
* num_kv_heads // world_size < 1

I suppose `num_heads` can be constant. Just don't make it a global constant. You can hard-code inside the test function and then pass it into the nn.Module __init__ function.

**Comment by greg-kwasniewski1** [comment]
@lucaslie  Implemented custom GQA_block in the `test_graph_sharding.py`. Now everything works nicely.

**Review by lucaslie — APPROVED** [review]
~@greg-kwasniewski1 , this looks great. I kicked off a dashboard run with your changes: https://gitlab-master.nvidia.com/ftp/infra/autodeploy-dashboard/-/pipelines/30068368~

~I think it's worth checking what happens and if all models pass as before. Let's monitor it and see what happens~

Nevermind, Let's get it merged and look at the dashboard tomorrow morning after it ran on the feature branch :)
